### PR TITLE
Add operation.with_directives decorator

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Adds a new decorator which add the appropriate directives along with registering the function as an operation `FlowProject.operation.with_directives`.

To accomplish this, we create a new class `OperationRegister`, that is specific to a `FlowProject` class like the preconditions and postconditions classes. We use the `__call__` magic method for the basic decorator.

## Motivation and Context

Addresses: #309

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [x] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: tests/test_project.py [hunk 1]
```diff
@@ -708,6 +708,21 @@ def b(job):
             not callable(value) for value in submit_job_operation.directives.values()
         )
 
+    def test_operation_with_directies(self):
```
Comment (bdice) [inline tests/test_project.py:711]:
```suggestion
    def test_operation_with_directives(self):
```

File: flow/project.py [hunk 2]
```diff
@@ -1388,6 +1391,131 @@ def copy_from(cls, *other_funcs):
 
         return post
 
+    @staticmethod
+    def _setup_operation_register_object(parent_class):
```
Comment (bdice) [inline flow/project.py:1394]: This name is awkward -- it's long and the words don't tell me very much. I would prefer `_setup_operation_class`, to mirror `_setup_preconditions_class` and `_setup_postconditions_class`.

File: flow/project.py [hunk 3]
```diff
@@ -1388,6 +1391,131 @@ def copy_from(cls, *other_funcs):
...
+                    # FlowGroupEntry(name=name, options="", aggregator=func._flow_aggregate)
+                )
+                if hasattr(func, "_flow_groups"):
+                    func._flow_groups.append(name)
+                else:
+                    func._flow_groups = [name]
+                return func
+
+            @classmethod
+            def with_directives(cls, directives, name=None):
```
Comment (bdice) [inline flow/project.py:1491]: I would prefer 
```suggestion
            def with_directives(cls, name, **directives):
```
as originally proposed in #309.

File: doc/api.rst [hunk 4]
```diff
@@ -58,8 +59,11 @@ The FlowProject
 .. autoclass:: FlowProject
     :show-inheritance:
     :members:
-    :exclude-members: pre,post
+    :exclude-members: pre,post,operation
 
+.. automethod:: flow.FlowProject.operation(func, name=None)
```
Comment (bdice) [inline doc/api.rst:64]: Do you have to specify the arguments, or can sphinx detect them automatically?

**Review by bdice — COMMENTED** [review]
@b-butler I was curious what this PR did, so I gave it a very quick read-through. I recognize it's early stage, but thought that some design feedback might be helpful.

File: doc/api.rst (hunk 4 above)
Comment (b-butler) [reply doc/api.rst:64]: Have to

File: flow/project.py (hunk 3 above)
Comment (b-butler) [reply flow/project.py:1491]: We mention keeping the same API as in `FlowGroup.with_directives` which uses a dictionary.

**Comment by b-butler** [comment]
Here is the documentation PR https://github.com/glotzerlab/signac-docs/pull/128

I also added tests for lines of code that code coverage was failing me for not testing even though I did not write those lines of code, in doing so I fixed a small bug.

**Review by bdice — CHANGES_REQUESTED** [review]
We should also modify the test files like `define_test_project.py` so that they use a mixture of the old and new syntax, at least until we deprecate/remove `@directives`.

File: flow/project.py [hunk 5]
```diff
@@ -1388,6 +1389,125 @@ def copy_from(cls, *other_funcs):
...
+                    raise ValueError(
+                        f"A group with name '{name}' is already registered."
+                    )
+
+                signature = inspect.signature(func)
+                for i, parameter_value in enumerate(signature.parameters.values()):
+                    if i > 0 and parameter_value.default is inspect.Parameter.empty:
+                        raise ValueError(
+                            "Only the first argument in an operation argument may not have "
+                            f"a default value! ({name})"
```
Comment (bdice) [inline flow/project.py:1466]: Is this a potential issue for aggregation? The user could specify:

```python
@FlowProject.operation
@aggregator.groupsof(2)
def my_operation(job1, job2):
    print(job1, job2)
```
and I'm also unsure of how this handles `*jobs`.

File: flow/project.py [hunk 6]
```diff
@@ -4184,6 +4231,12 @@ def foo(job):
         # TODO: Add aggregator argument and pass through to FlowGroupEntry.
         if name in cls._GROUP_NAMES:
             raise ValueError(f"Repeat definition of group with name '{name}'.")
+        if any(
+            name == operation_name for operation_name, _ in cls._OPERATION_FUNCTIONS
+        ):
+            raise ValueError(
+                f"Cannot create user group with name of existing operation {name}"
```
Comment (bdice) [inline flow/project.py:4238]: "User group" makes this sound like a UNIX authentication problem...
```suggestion
                f"Cannot create a group with the same name as the existing operation {name}"
```

File: tests/define_test_project.py [hunk 7]
```diff
@@ -53,20 +56,19 @@ def _need_to_fork(job):
...
 
 
-@_TestProject.operation
+@_TestProject.operation.with_directives({"ngpu": 1, "omp_num_threads": 1})
 @group2.with_directives(dict(omp_num_threads=4))
 @_TestProject.post.true("test3_true")
 @_TestProject.post.false("test3_false")
 @_TestProject.post.not_(lambda job: job.doc.test3_false)
-@flow.directives(ngpu=1, omp_num_threads=1)
+@flow.directives()
```
Comment (bdice) [inline tests/define_test_project.py:71]: Looks like this was left by mistake.
```suggestion
```

File: tests/test_project.py [hunk 8]
```diff
@@ -708,6 +734,38 @@ def b(job):
...
+            for next_op in project._next_operations([(job,)]):
+                assert next_op.directives["executable"] == "python3"
+                assert next_op.directives["nranks"] == 4
+            break
+
+    def test_old_directives_decorator(self):
+        class A(FlowProject):
+            pass
+
+        # add deprecation warning context manager in v0.15
```
Comment (bdice) [inline tests/test_project.py:756]:
```suggestion
        # TODO: Add deprecation warning context manager in v0.15
```

**Review by bdice — APPROVED** [review]
I have a few smaller requests / questions. I approve after those comments are addressed. @lyrivera @kidrahahjo You're both assigned to this PR, could one of you review? Thanks!

File: flow/project.py (hunk 5 above)
Comment (b-butler) [reply flow/project.py:1466]: This is also a current check (just copied and pasted code), so I don't think this PR should fix that.

**Comment by bdice** [comment]
@b-butler Can you resolve merge conflicts?

**Comment by b-butler** [comment]
@bdice Done.

**Review by kidrahahjo — APPROVED** [review]
I really like this implementation. These changes look good to me. @b-butler @bdice Is there a need to add documentation for this in signac-docs?

**Comment by b-butler** [comment]
@kidrahahjo there is a docs PR https://github.com/glotzerlab/signac-docs/pull/128

**Pull request merged**